RecoCTPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc

```diff
@@ -0,0 +1,357 @@
+
+    // increase chi^2
+    s2 += x_diff_norm*x_diff_norm + y_diff_norm*y_diff_norm;
+  }
+
+  return s2;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void ProtonReconstructionAlgorithm::reconstructFromMultiRP(const CTPPSLocalTrackLiteRefVector& tracks,
+                                                           reco::ForwardProtonCollection& output,
+                                                           const LHCInfo& lhcInfo, std::ostream& os) const
+{
+  if (!initialized_)
+    return;
+
+  // need at least two tracks
+  if (tracks.size() < 2)
+    return;
+
+  // make sure optics is available for all tracks
+  for (const auto &it : tracks) {
+    auto oit = m_rp_optics_.find(it->getRPId());
+    if (oit == m_rp_optics_.end())
+      throw cms::Exception("ProtonReconstructionAlgorithm") << "Optics data not available for RP " <<
+        it->getRPId() << ", i.e. " << CTPPSDetId(it->getRPId()) << ".";
+  }
+
+  // initial estimate of xi and th_x
+  double xi_init = 0., th_x_init = 0.;
+
+  if (useImprovedInitialEstimate_) {
+    double x_N = tracks[0]->getX()*1E-1,  // conversion: mm --> cm
+      x_F = tracks[1]->getX()*1E-1;
+
+    const RPOpticsData &i_N = m_rp_optics_.find(tracks[0]->getRPId())->second,
+      &i_F = m_rp_optics_.find(tracks[1]->getRPId())->second;
+
+    const double a = i_F.ch1*i_N.la1 - i_N.ch1*i_F.la1;
+    const double b = i_F.ch0*i_N.la1 - i_N.ch0*i_F.la1 + i_F.ch1*i_N.la0 - i_N.ch1*i_F.la0 + x_N*i_F.la1 - x_F*i_N.la1;
+    const double c = x_N*i_F.la0 - x_F*i_N.la0 + i_F.ch0*i_N.la0 - i_N.ch0*i_F.la0;
+    const double D = b*b - 4.*a*c;
+
+    xi_init = (-b + sqrt(D)) / 2. / a;
+    th_x_init = (x_N - i_N.ch0 - i_N.ch1 * xi_init) / (i_N.la0 + i_N.la1 * xi_init);
+  }
+  else {
+    double s_xi0 = 0., s_1 = 0.;
+    for (const auto &track : tracks) {
+      auto oit = m_rp_optics_.find(track->getRPId());
+      double xi = oit->second.s_xi_vs_x_d->Eval(track->getX()*1E-1 + oit->second.x0);  // conversion: mm --> cm
+
+      s_1 += 1.;
+      s_xi0 += xi;
+    }
+
+    xi_init = s_xi0 / s_1;
+  }
+
+  // initial estimate of th_y and vtx_y
+  double y[2], v_y[2], L_y[2];
+  unsigned int y_idx = 0;
+  for (const auto &track : tracks) {
+    if (y_idx >= 2)
+      continue;
+
+    auto oit = m_rp_optics_.find(track->getRPId());
+
+    y[y_idx] = track->getY()*1E-1 - oit->second.s_y_d_vs_xi->Eval(xi_init); // track y: mm --> cm
+    v_y[y_idx] = oit->second.s_v_y_vs_xi->Eval(xi_init);
+    L_y[y_idx] = oit->second.s_L_y_vs_xi->Eval(xi_init);
+
+    y_idx++;
+  }
+
+  double vtx_y_init = 0.;
+  double th_y_init = 0.;
+
+  if (fitVtxY_) {
+    const double det_y = v_y[0] * L_y[1] - L_y[0] * v_y[1];
+    vtx_y_init = (L_y[1] * y[0] - L_y[0] * y[1]) / det_y;
+    th_y_init = (v_y[0] * y[1] - v_y[1] * y[0]) / det_y;
+  }
+  else {
+    vtx_y_init = 0.;
+    th_y_init = (y[1]/L_y[1] + y[0]/L_y[0]) / 2.;
```

Comment:
static analysis suggests there may be a bug here.
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25495/33634/llvm-analysis/report-85a4a3.html#EndPath

Reply:
The same answer as here: https://github.com/cms-sw/cmssw/pull/25495#discussion_r266961724.